File: desktop/electron/services/chatService.ts
import { randomUUID } from 'node:crypto';
import type { BrowserWindow } from 'electron';
import type {
  ChatCompletionMessageParam,
  ChatCompletionTool,
} from 'openai/resources/chat/completions';
import type {
  AppSettings,
  ChatDoneEvent,
  ChatErrorEvent,
  ChatSendRequest,
  ChatTokenEvent,
  ServerProfile,
  WorkspaceOpEvent,
} from '../../../shared/types';
import { ProviderError, type LlmClient } from '../../../shared/llm/types';
import { parseActionTags } from '../../../shared/workspace/actionTags';
import {
  WorkspaceError,
  WorkspaceService,
  workspaceTools,
} from '../../../shared/workspace/workspaceService';
import { ConversationStore } from './conversationStore';
import { PersonaRegistry } from './personaRegistry';
import { ImageService } from './imageService';
import { OrchestratorService } from './orchestratorService';
import { emitModelStatus } from './modelStatus';

const MAX_TOOL_ROUNDS = 8;

export class ChatService {
  private abortControllers = new Map<string, AbortController>();
  private activeMessageId: string | null = null;
  private workspace = new WorkspaceService();
  private orchestrator: OrchestratorService;

  constructor(
    private getClientFor: (serverId: string | null) => LlmClient,
    private store: ConversationStore,
    private personas: PersonaRegistry,
    private images: ImageService,
    private getSettings: () => AppSettings,
    private getWindow: () => BrowserWindow | null,
  ) {
    this.orchestrator = new OrchestratorService(
      getClientFor,
      store,
      personas,
      getSettings,
      getWindow,
    );
  }

  private resolveProfile(serverId: string | null): ServerProfile | null {
    if (!serverId) return null;
    return this.getSettings().servers.find((server) => server.id === serverId) ?? null;
  }

  async send(
    request: ChatSendRequest,
  ): Promise<{ userMessageId: string; assistantMessageId: string }> {
    const existing = this.store.getConversation(request.conversationId);
    if (existing?.kind === 'orchestrator') {
      this.abortControllers.get(request.conversationId)?.abort();
      this.abortControllers.delete(request.conversationId);
      return this.orchestrator.send(request);
    }

    this.abortControllers.get(request.conversationId)?.abort();
    this.abortControllers.delete(request.conversationId);

    const conversation = existing;
    if (!conversation) {
      throw new ProviderError('unknown', 'Conversation not found');
    }

    const settings = this.getSettings();
    const profile = this.resolveProfile(conversation.serverId);
    const persona = this.personas.get(request.personaId) ?? this.personas.list()[0];
    const model = persona.defaultModel || conversation.model || settings.model;

    if (!model) {
      throw new ProviderError(
        'model_not_loaded',
        'No model selected. Choose a model in the top bar or Settings.',
      );
    }

    const userMessage = this.store.addMessage({
      conversationId: request.conversationId,
      role: 'user',
      content: request.content,
    });

    const assistantMessageId = randomUUID();
    this.activeMessageId = assistantMessageId;

    const history = this.store.getMessages(request.conversationId);
    const capped = history.slice(-Math.max(1, profile?.maxHistory ?? settings.maxHistory));
    const workspacePath = conversation.workspacePath;

    const systemParts = [persona.systemPrompt];
    if (workspacePath) {
      let tree: string;
      try {
        tree = this.workspace.buildTree(workspacePath);
      } catch (error) {
        tree = error instanceof Error ? error.message : String(error);
      }

      systemParts.push(
        [
          `You have a writable workspace folder bound to this chat: ${workspacePath}`,
          'You may inspect and modify files inside this folder only.',
          'Prefer tools when available. If tools are unavailable, emit exact XML actions:',
          '<list_dir path="." />',
          '<read_file path="relative/path.ext" />',
          '<write_file path="relative/path.ext">FULL FILE CONTENT</write_file>',
          '<delete_file path="relative/path.ext" />',
          '<generate_image path="images/out.png" prompt="a red circle" size="512x512" />',
          'After file work, give a short summary of what changed.',
          'Current workspace tree:',
          tree,
        ].join('\n'),
      );
    }

    const openaiMessages: ChatCompletionMessageParam[] = [
      { role: 'system', content: systemParts.join('\n\n') },
      ...capped
        .filter((message) => message.role === 'user' || message.role === 'assistant')
        .map((message) => ({
          role: message.role as 'user' | 'assistant',
          content: message.content,
        })),
    ];

    const win = this.getWindow();
    const myController = new AbortController();
    this.abortControllers.set(request.conversationId, myController);
    const client = this.getClientFor(conversation.serverId);
    let fullContent = '';

    try {
      await client.ensureModelLoaded(model, {
        signal: myController.signal,
        onStatus: (message) => emitModelStatus(this.getWindow, model, message),
      });

      if (workspacePath) {
        fullContent = await this.runWorkspaceAgent({
          workspacePath,
          model,
          messages: openaiMessages,
          conversationId: request.conversationId,
          messageId: assistantMessageId,
          signal: myController.signal,
          serverId: conversation.serverId,
        });
      } else {
        for await (const delta of client.streamChat(openaiMessages, model, myController.signal)) {
          fullContent += delta;
          this.emitToken(request.conversationId, assistantMessageId, delta);
        }
      }

      this.store.addMessage({
        id: assistantMessageId,
        conversationId: request.conversationId,
        role: 'assistant',
        content: fullContent,
        personaId: persona.id,
      });

      const doneEvent: ChatDoneEvent = {
        conversationId: request.conversationId,
        messageId: assistantMessageId,
        content: fullContent,
        personaId: persona.id,
      };
      win?.webContents.send('chat:done', doneEvent);

      return { userMessageId: userMessage.id, assistantMessageId };
    } catch (error) {
      const mapped =
        error instanceof ProviderError
          ? error
          : new ProviderError('unknown', error instanceof Error ? error.message : String(error));

      if (mapped.code !== 'cancelled' && fullContent) {
        this.store.addMessage({
          id: assistantMessageId,
          conversationId: request.conversationId,
          role: 'assistant',
          content: fullContent,
          personaId: persona.id,
        });
      }

      const errorEvent: ChatErrorEvent = {
        conversationId: request.conversationId,
        messageId: assistantMessageId,
        code: mapped.code,
        message: mapped.message,
      };
      win?.webContents.send('chat:error', errorEvent);

      if (mapped.code !== 'cancelled') {
        throw mapped;
      }

      return { userMessageId: userMessage.id, assistantMessageId };
    } finally {
      if (this.abortControllers.get(request.conversationId) === myController) {
        this.abortControllers.delete(request.conversationId);
      }
      if (this.activeMessageId === assistantMessageId) {
        this.activeMessageId = null;
      }
    }
  }

  private async runWorkspaceAgent(input: {
    workspacePath: string;
    model: string;
    messages: ChatCompletionMessageParam[];
    conversationId: string;
    messageId: string;
    signal: AbortSignal;
    serverId: string | null;
  }): Promise<string> {
    const client = this.getClientFor(input.serverId);
    const messages = [...input.messages];
    let finalText = '';
    let toolsEnabled = true;

    for (let round = 0; round < MAX_TOOL_ROUNDS; round += 1) {
      if (input.signal.aborted) {
        throw new ProviderError('cancelled', 'Generation cancelled');
      }

      let completion;
      try {
        completion = await client.completeChat(messages, input.model, {
          tools: toolsEnabled ? (workspaceTools as ChatCompletionTool[]) : undefined,
          signal: input.signal,
        });
      } catch (error) {
        if (toolsEnabled) {
          toolsEnabled = false;
          completion = await client.completeChat(messages, input.model, {
            signal: input.signal,
          });
        } else {
          throw error;
        }
      }

      const nativeCalls = completion.toolCalls;
      if (nativeCalls.length) {
        messages.push({
          role: 'assistant',
          content: completion.content || null,
          tool_calls: nativeCalls.map((call) => ({
            id: call.id,
            type: 'function' as const,
            function: { name: call.name, arguments: call.arguments },
          })),
        });

        for (const call of nativeCalls) {
          const result = await this.runToolAndEmit({
            workspacePath: input.workspacePath,
            conversationId: input.conversationId,
            messageId: input.messageId,
            name: call.name,
            rawArgs: call.arguments,
            serverId: input.serverId,
          });
          messages.push({
            role: 'tool',
            tool_call_id: call.id,
            content: result,
          });
        }
        continue;
      }

      const actions = parseActionTags(completion.content || '');
      if (actions.length) {
        messages.push({
          role: 'assistant',
          content: completion.content || '',
        });

        const results: string[] = [];
        for (const action of actions) {
          const result = await this.runToolAndEmit({
            workspacePath: input.workspacePath,
            conversationId: input.conversationId,
            messageId: input.messageId,
            name: action.name,
            rawArgs: JSON.stringify(action.args),
            serverId: input.serverId,
          });
          results.push(`[${action.name}] ${result}`);
        }

        messages.push({
          role: 'user',
          content:
            'Tool results:\n' +
            results.join('\n\n') +
            '\n\nContinue. If more file work is needed, emit more actions. Otherwise summarize for the user.',
        });
        continue;
      }

      finalText = (completion.content || '').trim();
      break;
    }

    if (!finalText) {
      finalText = 'Finished workspace operations.';
    }

    this.emitToken(input.conversationId, input.messageId, finalText);
    return finalText;
  }

  private async runToolAndEmit(input: {
    workspacePath: string;
    conversationId: string;
    messageId: string;
    name: string;
    rawArgs: string;
    serverId: string | null;
  }): Promise<string> {
    let args: Record<string, unknown>;
    try {
      args = JSON.parse(input.rawArgs || '{}') as Record<string, unknown>;
    } catch {
      args = {};
    }

    const relPath = String(
      args.path ?? (input.name === 'generate_image' ? 'images/generated.png' : '.'),
    );
    this.emitOp({
      conversationId: input.conversationId,
      messageId: input.messageId,
      op: input.name as WorkspaceOpEvent['op'],
      path: relPath,
      status: 'running',
    });

    const capturesCheckpoint = input.name === 'write_file' || input.name === 'delete_file';
    const previousContent = capturesCheckpoint
      ? this.workspace.tryReadFile(input.workspacePath, relPath)
      : null;

    try {
      let result: string;
      if (input.name === 'generate_image') {
        result = await this.images.generateForTool({
          prompt: String(args.prompt ?? ''),
          workspaceRoot: input.workspacePath,
          relativePath: relPath,
          size: args.size ? String(args.size) : undefined,
          conversationId: input.conversationId,
          serverId: input.serverId,
        });
      } else {
        result = this.workspace.executeTool(input.workspacePath, input.name, args);
      }

      const checkpointId = capturesCheckpoint
        ? this.store.addCheckpoint({
            conversationId: input.conversationId,
            relativePath: relPath,
            previousContent,
            previousExisted: previousContent !== null,
          }).id
        : undefined;

      this.emitOp({
        conversationId: input.conversationId,
        messageId: input.messageId,
        op: input.name as WorkspaceOpEvent['op'],
        path: relPath,
        status: 'ok',
        detail: result.slice(0, 240),
        checkpointId,
      });
      return result;
    } catch (error) {
      const message =
        error instanceof WorkspaceError || error instanceof Error ? error.message : String(error);
      this.emitOp({
        conversationId: input.conversationId,
        messageId: input.messageId,
        op: input.name as WorkspaceOpEvent['op'],
        path: relPath,
        status: 'error',
        detail: message,
      });
      return `Error: ${message}`;
    }
  }

  private emitToken(conversationId: string, messageId: string, delta: string): void {
    const tokenEvent: ChatTokenEvent = { conversationId, messageId, delta };
    this.getWindow()?.webContents.send('chat:token', tokenEvent);
  }

  private emitOp(event: WorkspaceOpEvent): void {
    this.getWindow()?.webContents.send('workspace:op', event);
  }

  cancel(conversationId: string): boolean {
    const orchCancelled = this.orchestrator.cancel(conversationId);
    const controller = this.abortControllers.get(conversationId);
    if (!controller) return orchCancelled;
    controller.abort();
    this.abortControllers.delete(conversationId);
    return true;
  }

  getActiveMessageId(): string | null {
    return this.activeMessageId ?? this.orchestrator.getActiveMessageId();
  }
}

File: desktop-java/src/main/java/com/multiagent/desktop/service/ChatService.java
package com.multiagent.desktop.service;

import com.multiagent.desktop.action.ActionApprover;
import com.multiagent.desktop.llm.CancellationToken;
import com.multiagent.desktop.llm.ChatRequestMessage;
import com.multiagent.desktop.llm.ErrorCode;
import com.multiagent.desktop.llm.LlmClient;
import com.multiagent.desktop.llm.ProviderException;
import com.multiagent.desktop.model.ChatMessage;
import com.multiagent.desktop.model.Conversation;
import com.multiagent.desktop.model.ImageAttachment;
import com.multiagent.desktop.model.MessageRole;
import com.multiagent.desktop.model.Persona;
import com.multiagent.desktop.persistence.ConversationStore;
import com.multiagent.desktop.workspace.WorkspaceService;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.function.Consumer;

/**
 * Chat turns for both plain conversations and workspace-bound ones, mirroring
 * desktop/electron/services/chatService.ts's send(). Plain chats stream tokens directly;
 * workspace-bound ones (Conversation.workspacePath set) delegate to ToolLoopRunner, which
 * isn't token-streamed - like the TS version, its whole answer arrives as one onToken call
 * once the tool-calling loop finishes. Per-conversation CancellationToken map means two
 * conversations can generate concurrently without one cancelling the other.
 */
public class ChatService {
    private final ConversationStore store;
    private final WorkspaceService workspace = new WorkspaceService();
    private final ToolLoopRunner toolLoopRunner;
    private final Map<String, CancellationToken> activeTokens = new ConcurrentHashMap<>();
    private final ExecutorService executor = Executors.newCachedThreadPool(runnable -> {
        Thread thread = new Thread(runnable, "chat-service-worker");
        thread.setDaemon(true);
        return thread;
    });

    public ChatService(ConversationStore store) {
        this.store = store;
        this.toolLoopRunner = new ToolLoopRunner(store);
    }

    /** Installs the "ask before writing/deleting/renaming a file" gate - see ToolLoopRunner's javadoc for why a null approver auto-approves (tests only). */
    public void setActionApprover(ActionApprover approver) {
        toolLoopRunner.setActionApprover(approver);
    }

    public interface Listener {
        void onToken(String conversationId, String messageId, String delta);

        void onDone(String conversationId, ChatMessage message);

        void onError(String conversationId, String messageId, ErrorCode code, String message);

        /** status: "running" | "ok" | "error". checkpointId is non-null only for a successful write_file/delete_file. No-op default for callers that don't care (e.g. tests). */
        default void onWorkspaceOp(String conversationId, String messageId, String op, String path,
                                    String status, String detail, String checkpointId) {
        }

        /** Orchestrator-only: phase is "planning" | "specialist" | "synthesizing" | "done". */
        default void onStep(String conversationId, String phase, String personaId, String label) {
        }

        /** Model load/availability progress before generation starts, e.g. "Loading qwen2.5-coder...". No-op default. */
        default void onModelStatus(String conversationId, String status) {
        }

        /** Orchestrator-only: fired after each specialist note is persisted, so the UI can show it before the final synthesis arrives. */
        default void onMessagesUpdated(String conversationId) {
        }
    }

    /** Persists the user message, then generates+persists the assistant reply on a background thread. */
    public void send(LlmClient client, Conversation conversation, String content, Persona persona,
                      String model, String visionModel, int maxHistory, int contextTokens,
                      ImageAttachment image, Listener listener) {
        String persistedContent = image != null
                ? (content.isBlank() ? "" : content + "\n\n") + "[🖼️ " + image.name() + "]"
                : content;
        store.addMessage(conversation.getId(), MessageRole.USER, persistedContent, persona.getId());

        List<ChatMessage> history = store.getMessages(conversation.getId());
        List<ChatRequestMessage> messages = new ArrayList<>();
        boolean visionEnabled = visionModel != null && !visionModel.isBlank();
        messages.add(ChatRequestMessage.system(buildSystemPrompt(persona, conversation, visionEnabled)));
        int from = Math.max(0, history.size() - maxHistory);
        for (ChatMessage message : history.subList(from, history.size())) {
            if (message.getRole() == MessageRole.USER || message.getRole() == MessageRole.ASSISTANT) {
                messages.add(ChatRequestMessage.of(message.getRole(), message.getContent()));
            }
        }

        // Inline mode: when the chat model IS the vision model (it can see for itself), attach
        // the image to this turn's user message instead of the transcribe-then-splice pre-pass.
        boolean inlineImage = image != null && visionEnabled && visionModel.equals(model);
        if (inlineImage) {
            messages.set(messages.size() - 1, ChatRequestMessage.userWithImage(content, image));
        }

        CancellationToken token = new CancellationToken();
        activeTokens.put(conversation.getId(), token);
        String assistantMessageId = UUID.randomUUID().toString();
        StringBuilder buffer = new StringBuilder();
        String workspacePath = conversation.getWorkspacePath();

        executor.submit(() -> {
            try {
                if (image != null && !inlineImage) {
                    describeAttachedImage(client, visionModel, image, messages,
                            status -> listener.onModelStatus(conversation.getId(), status), token);
                }

                client.ensureModelLoaded(model,
                        status -> listener.onModelStatus(conversation.getId(), status), token);

                // Fit the request to the server's context window: drop the oldest turns until
                // the estimate leaves room for the reply, and cap generation to that room so
                // the server can't overflow either. ctx <= 0 ⇒ unknown, keep maxHistory only.
                int maxTokens = 0;
                if (contextTokens > 0) {
                    int reserve = Math.max(512, Math.min(contextTokens / 4, 4096));
                    int budget = contextTokens - reserve;
                    while (messages.size() > 2 && TokenEstimate.estimateMessages(messages) > budget) {
                        messages.remove(1); // index 0 is the system prompt; keep it + the final turn
                    }
                    maxTokens = reserve;
                }

                if (workspacePath != null && !workspacePath.isBlank()) {
                    String finalText = toolLoopRunner.run(client, model, visionModel, maxTokens, messages,
                            workspacePath, conversation.getId(), token,
                            (op, path, status, detail, checkpointId) -> listener.onWorkspaceOp(
                                    conversation.getId(), assistantMessageId, op, path, status, detail, checkpointId));
                    buffer.append(finalText);
                    listener.onToken(conversation.getId(), assistantMessageId, finalText);
                } else {
                    client.streamChat(messages, model, maxTokens, delta -> {
                        buffer.append(delta);
                        listener.onToken(conversation.getId(), assistantMessageId, delta);
                    }, token);
                }

                ChatMessage saved = store.addMessage(conversation.getId(), MessageRole.ASSISTANT,
                        buffer.toString(), persona.getId());
                listener.onDone(conversation.getId(), saved);
            } catch (ProviderException e) {
                listener.onError(conversation.getId(), assistantMessageId, e.code(), e.getMessage());
            } catch (Exception e) {
                listener.onError(conversation.getId(), assistantMessageId, ErrorCode.UNKNOWN,
                        e.getMessage() != null ? e.getMessage() : e.toString());
            } finally {
                activeTokens.remove(conversation.getId());
            }
        });
    }

    /**
     * "Dash of B": before the real turn runs, transcribe the attached image once with the
     * vision model and splice the description in just ahead of the user's message, so even a
     * model that never calls describe_image has the image's content. Failures degrade to a
     * short note rather than aborting the turn. Runs on the send() background thread.
     */
    private void describeAttachedImage(LlmClient client, String visionModel, ImageAttachment image,
                                        List<ChatRequestMessage> messages, Consumer<String> onStatus,
                                        CancellationToken token) {
        int insertAt = Math.max(1, messages.size() - 1); // before the just-persisted user turn
        if (visionModel == null || visionModel.isBlank()) {
            messages.add(insertAt, ChatRequestMessage.user("[An image \"" + image.name()
                    + "\" was attached, but this server has no vision model configured, so it could not be read.]"));
            return;
        }
        try {
            onStatus.accept("Looking at " + image.name() + "...");
            client.ensureModelLoaded(visionModel, onStatus, token);
            String description = client.describeImage(visionModel,
                    "Describe this image in full detail, including any text, code, diagrams, tables, and UI elements.",
                    image.bytes(), image.mimeType(), token);
            String note = VisionResponses.looksLikeRefusal(description)
                    ? "[The vision model could not read the attached image \"" + image.name()
                        + "\". Tell the user it couldn't be processed; do not attempt to describe it.]"
                    : "[Image \"" + image.name() + "\" attached. Vision model description:]\n\n" + description.strip();
            messages.add(insertAt, ChatRequestMessage.user(note));
        } catch (RuntimeException e) {
            messages.add(insertAt, ChatRequestMessage.user("[Could not read the attached image \""
                    + image.name() + "\": " + (e.getMessage() != null ? e.getMessage() : e.toString()) + "]"));
        }
    }

    /** Persona prompt, plus (when workspace-bound) the directory tree and tool-usage instructions. */
    private String buildSystemPrompt(Persona persona, Conversation conversation, boolean visionEnabled) {
        List<String> parts = new ArrayList<>();
        parts.add(persona.getSystemPrompt());

        String workspacePath = conversation.getWorkspacePath();
        if (workspacePath != null && !workspacePath.isBlank()) {
            String tree;
            try {
                tree = workspace.buildTree(workspacePath);
            } catch (RuntimeException e) {
                tree = e.getMessage();
            }
            List<String> lines = new ArrayList<>(List.of(
                    "You have a writable workspace folder bound to this chat: " + workspacePath,
                    "You may inspect and modify files inside this folder only.",
                    "This folder IS the project root. Scaffold a new project directly into it (write `Cargo.toml`, "
                            + "`package.json`, `src/main.rs`, … at the top level) - do NOT create a subfolder named "
                            + "after the project, and do NOT run `cargo new` / `npm init <name>` style commands that "
                            + "would nest one.",
                    "Prefer tools when available. If tools are unavailable, emit exact XML actions:",
                    "<list_dir path=\".\" />",
                    "<read_file path=\"relative/path.ext\" />  (add offset=\"1\" limit=\"200\" to read only a line range)",
                    "<search_file path=\"relative/path.ext\" pattern=\"text\" ignore_case=\"true\" context=\"2\" />  "
                            + "(grep a large file for matching lines + line numbers; use before read_file on anything big)",
                    "<write_file path=\"relative/path.ext\">FULL FILE CONTENT</write_file>",
                    "<delete_file path=\"relative/path.ext\" />",
                    "<rename_file path=\"relative/old.ext\" newPath=\"relative/new.ext\" />",
                    "<run_command command=\"npm run build\" timeout_seconds=\"120\" />  "
                            + "(one build/test/lint command; put the whole line in command=, no shell/pipes; "
                            + "add cwd=\"subdir\" only for a monorepo)",
                    "<generate_image path=\"images/out.png\" prompt=\"a red circle\" size=\"512x512\" />"));

            if (visionEnabled) {
                lines.add("<describe_image path=\"relative/img.png\" question=\"...\" />  "
                        + "(ask the configured vision model about a .png/.jpg/.gif/.webp image in the workspace)");
            }

            if (isGitRepo(workspacePath)) {
                lines.add("This workspace is a git repository. You also have git tools (XML forms shown for fallback):");
                lines.add("<git_status />");
                lines.add("<git_diff patch=\"true\" staged=\"true\" path=\"relative/path.ext\" />  (omit patch for a diffstat only)");
                lines.add("<git_log count=\"15\" />");
                lines.add("<git_show ref=\"HEAD\" patch=\"true\" />");
                lines.add("<git_branch />");
                lines.add("<git_add path=\"relative/path.ext\" />");
                lines.add("<git_commit message=\"Short summary of the change\" all=\"true\" />");
            }

            lines.add("You also have run_command (a native tool): run ONE build/test/lint/run command with the "
                    + "project's own toolchain (read the project files first to know which - npm, cargo, mvn, "
                    + "dotnet, pytest, ./gradlew, …). No shell, no pipes, no chaining. It runs in the workspace "
                    + "root by default; pass cwd only for a monorepo subdirectory. A non-zero exit code comes "
                    + "back as text - read it and fix the cause (a \"could not find Cargo.toml/package.json\" "
                    + "means the project files aren't where you ran it - check with list_dir).");
            lines.add("Changes to files, git_add/git_commit and run_command require the user's approval and may be "
                    + "declined - if a tool result says the user declined, respect that and don't retry the same action.");
            lines.add("After file work, give a short summary of what changed.");
            lines.add("Workspace tree (top levels only; a trailing \"…\" means use list_dir to see inside):");
            lines.add(tree);
            parts.add(String.join("\n", lines));
        }
        return String.join("\n\n", parts);
    }

    /** A cheap ".git present?" check so the prompt only advertises git tools when they'd actually work. */
    private boolean isGitRepo(String workspacePath) {
        try {
            Path dotGit = Path.of(workspacePath).resolve(".git");
            return Files.isDirectory(dotGit) || Files.isRegularFile(dotGit);
        } catch (RuntimeException e) {
            return false;
        }
    }

    public void cancel(String conversationId) {
        CancellationToken token = activeTokens.get(conversationId);
        if (token != null) {
            token.cancel();
        }
    }

    /** Whether this specific conversation has a generation in flight right now - lets the UI show Stop/Send per chat rather than globally. */
    public boolean isActive(String conversationId) {
        return activeTokens.containsKey(conversationId);
    }

    public void shutdown() {
        executor.shutdownNow();
    }
}
